### PR TITLE
Add explicit support for BuildCraft, CoFH wrenches

### PR DESCRIPTION
**Pull request by firenoo**
Uses AE2's integration system to add support for Thermal Expansion wrenches
* Refactor `ToolQuartzWrench` and `ToolNetworkTool`, move BC optional interface `IToolWrench` to `IAEWrench`
* Adds `IToolHammer` to `IAEWrench`

Fixes https://github.com/GTNewHorizons/Applied-Energistics-2-Unofficial/issues/395

### CODE REVIEW

**Comment by miozune** [comment]
I don't understand the issue, ThermalExpansion Crescent Hammer doesn't let me pick up the cable without BC, even before my patch https://github.com/GTNewHorizons/Applied-Energistics-2-Unofficial/pull/385

**Comment by firenoo** [comment]
Since your code most recently touched wrench code, I assumed it had to do with the issue. Apologies if I offended you.
But even so, it's a bug with the way it is implemented, and id like to fix that. If BC is not loaded, then it will definitely not work since AE only knows about the crescent hammer's wrenching ability from IToolWrench, if cofh's source is trustable.

**Comment by miozune** [comment]
We can use `IToolHammer` from CoFH Core

And no, I don't feel offended. Everyone makes mistakes 😉

**Comment by firenoo** [comment]
I wanted to avoid having to make multiple optional deps.
 but I see now that the wrenchers map will fill up with random item entries so that'll need a rewrite anyways.
Sadly we can't put the onus on cofh to provide compat :( so maybe using IToolHammer is right (if the hammer actually extends it)

**Comment by Alastors** [comment]
Optional dependency > hard dependency when applicable

**Comment by firenoo** [comment]
we are already talking about optional deps here, just how to implement it 😉

**Comment by miozune** [comment]
It's not hard, `cofhloaded && item instanceof IToolHammer`, just like I did with `IToolWrench`
You can check code for crescent hammer  [here](https://github.com/CoFH/ThermalExpansion-1.12-Legacy/blob/1.7.10/src/main/java/cofh/thermalexpansion/item/tool/ItemWrench.java)

**Comment by firenoo** [comment]
Don't worry, I am no stranger to optional dep implementation. I meant that I don't necessarily agree with the previous implementation (both yours and mine) now that I've looked closer at how AE2 does integration. To maintain consistency with the rest of the mod - at least until we refactor that part (if we ever want to) - its probabaly better to use these annotations: https://github.com/GTNewHorizons/Applied-Energistics-2-Unofficial/blob/master/src/main/java/appeng/transformer/annotations/Integration.java and create a new IntegrationType for CoFHCore/ThermalExpansion (or rather, re-add it since it was removed at some point)

**Comment by miozune** [comment]
Isn't `@Integration` basically the same as `@Optional`? But `IIntegrationModule` implementation sounds good

**Comment by glowredman** [comment]
The problem with checking the modid to decide whether to enable/disable the compat is, that the APIs can (and often are) shaded in other mods. "CoolWrenchMod" might shade both the BC API and the CoFH API and expect the Interaction with AE to work, even though neither BuildCraft or CoFHCore are present. IIRC FML provides a method to check if an API is loaded, so maybe it is possible to use that.

**Comment by glowredman** [comment]
Just checked, it is `ModAPIManager.hasAPI(String)`

**Comment by miozune** [comment]
AE2 `IIntegrationModule` already checks that, see `IntegrationNode`

**Comment by firenoo** [comment]
@miozune implemented. lmk if it looks better

File: dependencies.gradle [hunk 1]
```diff
@@ -11,7 +11,9 @@ dependencies {
     compileOnly('com.github.GTNewHorizons:OpenComputers:1.9.17-GTNH:api') {transitive = false}
     compileOnly('com.github.GTNewHorizons:waila:1.6.0:dev') {transitive = false}
     compileOnly('com.github.GTNewHorizons:Railcraft:9.15.0:api') {transitive = false}
-
+    // Uncomment to add thermal expansion + foundation for testing
+//    runtimeOnlyNonPublishable('curse.maven:thermal-foundation-222880:2388753')
+//    runtimeOnlyNonPublishable('curse.maven:thermal-expansion-69163:2388759')
```
Comment (miozune) [inline dependencies.gradle:16]: nit: maybe move this to below compileOnly things?

File: src/main/java/appeng/api/implementations/items/IAEWrench.java [hunk 2]
```diff
@@ -16,10 +16,19 @@
...
+
 /**
- * Implemented on AE's wrench(s) as a substitute for if BC's API is not available.
+ * Implemented on AE's wrench(s), automatically pulls in CoFH and BC's interfaces if present
  */
-public interface IAEWrench {
+@InterfaceList(
+        value = { @Interface(iface = "cofh.api.item.IToolHammer", iname = IntegrationType.CoFHWrench),
+                @Interface(iface = "buildcraft.api.tools.IToolWrench", iname = IntegrationType.BuildCraftCore) })
+public interface IAEWrench extends IToolHammer, IToolWrench {
```
Comment (miozune) [inline src/main/java/appeng/api/implementations/items/IAEWrench.java:31]: This is probably not a good way to implement tools from other mods, as
- This is technically a breaking change to API. e.g. OC implements `IAEWrench` but also implements `IToolWrench` and `IToolHammer`, so it's actually not a breaking change for this mod https://github.com/GTNewHorizons/OpenComputers/blob/master/src/main/scala/li/cil/oc/common/item/Wrench.scala#L17 But we can't know what mod outside of NH might be implementing it.
- No other classes in `api` package uses `@Interface`. With this change we enforce mods compiling with AE2 api to have CoFH and BC in dev env if they have reference to `IAEWrench`.

File: src/main/java/appeng/integration/IntegrationType.java [hunk 3]
```diff
@@ -62,7 +63,7 @@ public enum IntegrationType {
 
     Chisel(IntegrationSide.BOTH, "Chisel", "chisel"),
 
-    Jabba(IntegrationSide.BOTH, "Jabba", "JABBA");
+    Jabba(IntegrationSide.BOTH, "Jabba", "JABBA"),;
```
Comment (miozune) [inline src/main/java/appeng/integration/IntegrationType.java:66]: nit: unnecessary comma, I guess you tried to put semicolon in new line but spotless declined...

**Review by miozune — COMMENTED** [review]
`PartP2PTunnelStatic` still references raw `IToolWrench`

File: src/main/java/appeng/api/implementations/items/IAEWrench.java (hunk 2 above)
Comment (firenoo) [reply src/main/java/appeng/api/implementations/items/IAEWrench.java:31]: I suppose so, was not really thinking too far ahead, but I don't think your points are too worrysome:
1) can use default methods
2) These will be compileOnly references. we can switch to compileOnlyApi and dependent mods should still be fine should they switch to the new api version

Comment (miozune) [reply src/main/java/appeng/api/implementations/items/IAEWrench.java:31]: 1. Sure, it'll work
2. Uhh I don't like it, it makes addons to pull unnecessary deps, but I understand it's trade-off with code redundancy. It's ok, BC and CoFH are common mods in 1.7.10 modding scene.

File: src/main/java/appeng/integration/IntegrationType.java (hunk 3 above)
Comment (firenoo) [reply src/main/java/appeng/integration/IntegrationType.java:66]: spotless has been getting on my nerves lately for debugging/hotswap, i disable it for dev and reenable only at the end so i missed this

File: dependencies.gradle [hunk 4]
```diff
@@ -3,21 +3,24 @@
...
-
     compileOnly('net.industrial-craft:industrialcraft-2:2.2.828-experimental:api')
     compileOnly('curse.maven:minefactory-reloaded-66672:2366150')
     compileOnly('pneumaticCraft:PneumaticCraft-1.7.10:1.12.7-152:api') {transitive = false}
     compileOnly('curse.maven:better-storage-232919:2731636')
     compileOnly('api:immibis:1')
 
+    // Uncomment to add thermal expansion + foundation for testing
+    runtimeOnlyNonPublishable('curse.maven:thermal-foundation-222880:2388753')
+    runtimeOnlyNonPublishable('curse.maven:thermal-expansion-69163:2388759')
```
Comment (miozune) [inline dependencies.gradle:22]: Forgot to turn these off?

File: dependencies.gradle [hunk 5]
```diff
@@ -3,21 +3,24 @@
 dependencies {
     implementation('com.github.GTNewHorizons:NotEnoughItems:2.4.3-GTNH:dev')
     implementation('curse.maven:cofh-core-69162:2388751')
-    compileOnly('com.github.GTNewHorizons:BuildCraft:7.1.36:dev') {transitive = false}
+    implementation('com.github.GTNewHorizons:BuildCraft:7.1.36:dev') {transitive = false}
```
Comment (miozune) [inline dependencies.gradle:6]: Same here

Comment (firenoo) [reply dependencies.gradle]: yep

**Review by miozune — APPROVED** [review]

**Pull request merged**